consent/strategy_oauth_test.go

```diff
@@ -1108,6 +1108,133 @@ func TestStrategyLoginConsentNext(t *testing.T) {
 	})
 }
 
+func TestStrategyDeviceLoginConsent(t *testing.T) {
+	ctx := context.Background()
+	reg := testhelpers.NewMockedRegistry(t, &contextx.Default{})
+	reg.Config().MustSet(ctx, config.KeyAccessTokenStrategy, "opaque")
+	reg.Config().MustSet(ctx, config.KeyConsentRequestMaxAge, time.Hour)
+	reg.Config().MustSet(ctx, config.KeyConsentRequestMaxAge, time.Hour)
+	reg.Config().MustSet(ctx, config.KeyScopeStrategy, "exact")
+	reg.Config().MustSet(ctx, config.KeySubjectTypesSupported, []string{"pairwise", "public"})
+	reg.Config().MustSet(ctx, config.KeySubjectIdentifierAlgorithmSalt, "76d5d2bf-747f-4592-9fbd-d2b895a54b3a")
+
+	publicTS, adminTS := testhelpers.NewOAuth2Server(ctx, t, reg)
+	adminClient := hydra.NewAPIClient(hydra.NewConfiguration())
+	adminClient.GetConfig().Servers = hydra.ServerConfigurations{{URL: adminTS.URL}}
+
+	oauth2Config := func(t *testing.T, c *client.Client) *oauth2.Config {
+		return &oauth2.Config{
+			ClientID:     c.GetID(),
+			ClientSecret: c.Secret,
+			Endpoint: oauth2.Endpoint{
+				DeviceAuthURL: publicTS.URL + "/oauth2/device/auth",
+				TokenURL:      publicTS.URL + "/oauth2/token",
+				AuthStyle:     oauth2.AuthStyleInHeader,
+			},
+		}
+	}
+
+	acceptDeviceHandler := func(t *testing.T) http.HandlerFunc {
+		return checkAndAcceptDeviceHandler(t, adminClient)
+	}
+
+	acceptLoginHandler := func(t *testing.T, subject string, payload *hydra.AcceptOAuth2LoginRequest) http.HandlerFunc {
+		return checkAndAcceptLoginHandler(t, adminClient, subject, func(*testing.T, *hydra.OAuth2LoginRequest, error) hydra.AcceptOAuth2LoginRequest {
+			if payload == nil {
+				return hydra.AcceptOAuth2LoginRequest{}
+			}
+			return *payload
+		})
+	}
+
+	acceptConsentHandler := func(t *testing.T, payload *hydra.AcceptOAuth2ConsentRequest) http.HandlerFunc {
+		return checkAndAcceptConsentHandler(t, adminClient, func(*testing.T, *hydra.OAuth2ConsentRequest, error) hydra.AcceptOAuth2ConsentRequest {
+			if payload == nil {
+				return hydra.AcceptOAuth2ConsentRequest{}
+			}
+			return *payload
+		})
+	}
+
+	createDefaultClient := func(t *testing.T) *client.Client {
+		c := &client.Client{GrantTypes: []string{"urn:ietf:params:oauth:grant-type:device_code"}}
+		return createClient(t, reg, c)
+	}
+	t.Run("case=should pass if both login and consent are granted and check remember flows as well as various payloads", func(t *testing.T) {
```

Comment:
Can you please expand this test suite?

1. Refresh token rotation should work
2. Add test to ensure only granted scope is actually added/used (e.g. omit openid to remove id token result from the request, or offline_access)
3. Re-use detection works (so doing the same request twice fails)

Reply:
I see that you may have been covering this in the oauth2 package. Please confirm :)

Reply:
ping @nsklikas

Reply:
Sorry, I thought I had answered but it seems my comment was never submitted. Yes they should all be covered in https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go:
1. https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go#L539
2. https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go#L461 and https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go#L487
3. https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go#L736 and https://github.com/nsklikas/hydra/blob/canonical-master/oauth2/oauth2_device_code_test.go#L780